pkg/pool-weighted/contracts/WeightedPool.sol

```diff
@@ -301,7 +327,15 @@ contract WeightedPool is BaseWeightedPool, WeightedPoolProtocolFees {
         WeightedPoolProtocolFees._updatePostJoinExit(postJoinExitInvariant);
     }
 
-    function _beforeProtocolFeeCacheUpdate() internal override {
+    /**
+     * @dev This function will revert when called within a Vault context (i.e. in the middle of a join or an exit).
+     *
+     * This function depends on the invariant value, which may be calculated incorrectly in the middle of a join or
+     * an exit because the state of the pool could be out of sync with the state of the vault. This makes the function
+     * unsafe to call in such contexts, and hence it is protected.
+     * See https://forum.balancer.fi/t/reentrancy-vulnerability-scope-expanded/4345 for reference.
+     */
+    function _beforeProtocolFeeCacheUpdate() internal override whenNotInVaultContext {
```

Comment:
I'm not sure I agree with this patch. According to our mental model, the problematic function is `updateProtocolFeePercentageCache`, not this one. This function is dangerous and the caller must take care, but the caller here is the external function, and hence that should receive the fix.

Note that it doesn't matter much here since this is the only occurance, but still.

Reply:
This is just for the patch.
The problem with the external function is that it `ProtocolFeeCache` does not know about the vault, so it'll need a reference to it. And we wanted to keep the patch as concise as possible.

The plan is to add that reference in master for a more definitive fix and for better maintainability in future versions.

Reply:
I've added some extra notes in the external functions in https://github.com/balancer-labs/balancer-v2-monorepo/pull/2205/commits/9f56af4c2319ca773bcade4da640681a6978faa8; PTAL.

Note again that these are only applicable to this patch in particular, since they actually refer to weighted pools even though the utils may be used in master for other pool types. But I think it makes sense for this particular deployment.